src/app/components/settings/settings.rs

```diff
@@ -0,0 +1,271 @@
+    {
+        let pref_window = self.upcast_ref::<libadwaita::PreferencesWindow>();
+        let window = pref_window.upcast_ref::<libadwaita::Window>();
+
+        window.connect_close_request(
+            clone!(@weak self as _self => @default-return gtk::Inhibit(false), move |_| {
+                let widget = imp::SettingsWindow::from_instance(&_self);
+
+                let bitrate = match widget.player_bitrate.selected() {
+                    0 => Bitrate::Bitrate96,
+                    1 => Bitrate::Bitrate160,
+                    2 => Bitrate::Bitrate320,
+                    _ => unreachable!(),
+                };
+                let alsa_device = widget.alsa_device.text();
+                let backend = match widget.audio_backend.selected() {
+                    0 => AudioBackend::PulseAudio,
+                    1 => AudioBackend::Alsa(alsa_device.as_str().into()),
+                    _ => unreachable!(),
+                };
+                let ap_port = widget.ap_port.text().parse().ok().and_then(|port| match port {
+                    0 => None,
+                    port => Some(port),
+                });
+                let prefers_dark_theme = widget.theme.selected() == 1;
+
+                let player_settings = SpotifyPlayerSettings {
+                    bitrate,
+                    backend,
+                    ap_port,
+                };
+                let window_geometry = WindowGeometry::new_from_gsettings();
+                let settings = SpotSettings {
+                    prefers_dark_theme,
+                    player_settings,
+                    window: window_geometry,
+                };
+
+                on_close(settings);
+                gtk::Inhibit(false)
+            }),
+        );
+    }
+}
+
+pub struct Settings {
+    parent: gtk::Window,
+    settings_window: SettingsWindow,
+    settings: Rc<RefCell<SpotSettings>>,
+}
+
+impl Settings {
+    pub fn new(parent: gtk::Window, model: SettingsModel) -> Self {
+        let settings_window = SettingsWindow::new();
+        let settings = Rc::new(RefCell::new(
+            SpotSettings::new_from_gsettings().unwrap_or_default(),
+        ));
+
+        settings_window.connect_close(clone!(@weak settings => move |new_settings| {
+            if settings.borrow().player_settings != new_settings.player_settings {
+                model.set_player_settings(new_settings.player_settings);
+            }
+        }));
+
+        settings_window.connect_show(clone!(@weak settings => move |_| {
+            settings.replace(SpotSettings::new_from_gsettings().unwrap_or_default());
+        }));
+
+        Self {
+            parent,
+            settings_window,
+            settings,
+        }
+    }
+
+    fn window(&self) -> &libadwaita::Window {
+        let pref_window = self
+            .settings_window
+            .upcast_ref::<libadwaita::PreferencesWindow>();
+        pref_window.upcast_ref::<libadwaita::Window>()
```

Comment:
think you can upcast_ref to a window in one go, no need to cast to a preferences window first